xwiki-platform-core/xwiki-platform-query/xwiki-platform-query-xwql/src/main/java/org/xwiki/query/xwql/internal/hql/PropertyPrinter.java

```diff
@@ -40,14 +42,23 @@ void print(PropertyInfo prop, Printer printer) throws Exception
             if (className != null) {
                 prop.alias = printer.getContext().getAliasGenerator().generate(prop.object.alias + "_" + prop.name);
                 printer.from.append(", ").append(className).append(" as ").append(prop.alias);
+                // Join the property value for lists
+                if (className.endsWith("DBStringListProperty")) {
+                    printer.from.append(" join ").append(prop.alias).append(".").append(prop.getValueField())
+                        .append(" ").append(prop.alias + prop.getValueField());
+                }
                 printer.where.append(" and ").append(prop.alias).append(".id.id=").append(prop.object.alias)
                     .append(".id").append(" and ").append(prop.alias).append(".id.name").append("='").append(prop.name)
                     .append("'");
                 // rewrite nodes
                 for (PPath p : prop.locations) {
                     String s = prop.alias + "." + prop.getValueField();
-                    if (className.endsWith("DBStringListProperty") && p.parent() instanceof PSelectExpression) {
-                        s = "elements(" + s + ")";
+                    // Use the joined property value only in 'select', 'group by' and 'order by' statements
+                    if (className.endsWith("DBStringListProperty") && (
+                        p.parent() instanceof PSelectExpression ||
+                        p.parent().parent() instanceof POrderbyItem ||
```

Comment:
can this lead to a NPE? is it possible that `p.parent()` or `p.parent().parent()`  is null?

Reply:
There was no null checks for `p.parent()` before but it might be a good idea to write them, just in case.

Reply:
it's most likely that it's not needed because there's always an expression like SELECT or UPDATE, INSERT, etc.

I was more concerned with ` p.parent().parent()`. Does it mean that if the parent is a SELECT then there's always a parent of the SELECT? I'm not sure what this means.

Reply:
Actually we don't need to check nullity for `p.parent().parent()` as it won't create any issue. `null instanceof Object` will never create a NPE, it will just return `false`.

Reply:
ok wasn't sure that `null instanceof XXX` would fail or return false. Cool.